Slot mappings documentation update (#10011)
RasaHQ/rasa: docs/docs/migration-guide.mdx

However, the previous behavior can still be enforced by setting
`ranking_length=10` and `renormalize_confidences=True`.

### Slot Mappings

As of version 3.0, there is a single explicit mechanism for slot filling enabled by defining slot mappings for each slot

> in the `slots` section of the domain file. This approach allows slots to be kept up to date over the course of a conversation, and removes duplicated effort in
> mapping the same slots in multiple forms.

Reviewer: ```suggestion
in the `slots` section of the domain file. This approach keeps slots up to date over the course of a conversation, and removes duplicated effort in
mapping the same slots in multiple forms.
```

This new mechanism replaces the implicit slot setting via auto-fill of slots with entities of the same name.
The `auto_fill` key in the domain is no longer available, as well as the `auto_fill` parameter in the constructor of
the `Slot` class.

While forms continue to request the next slot, slot extraction is now delegated to the default

> action `action_extract_slots`. This action runs in the background automatically after each user turn.

Reviewer: link to default action docs
```suggestion
action [`action_extract_slots`](link). This action runs in the background automatically after each user turn.
```

> This approach allows slots to be kept up to date over the course of a conversation, and removes duplicated effort in

Reviewer 2: Paragraph starting on line 142 can probably go in the first paragraph (makes it easier to find out what changed and why if you're only skimming/reading part way)

> While forms continue to request the next slot, the extraction of slots is now delegated exclusively to the new default
> action `action_extract_slots`. This action runs in the background after each user turn without it being required to be
> triggered manually.

Reviewer: ```suggestion
While forms continue to request the next slot, slot extraction is now delegated to the default
action `action_extract_slots`. This action runs in the background automatically after each user turn. Like `action_listen`, it should not be included in stories.
```

Reviewer: Or some other phrase to indicate that neither needs to nor should ever be run manually

> You are now required to add the `mappings` key for each slot in the 3.0 domain format, including any slots with custom
> mappings. The keys used to define predefined mappings in 2.0 were maintained. Additionally, you can define slots with
> custom mappings implemented in a custom action, like so:

Reviewer: What about slots that really don't have mappings? e.g. they are just there to be filled in the case that a custom action wants to set them. 

```suggestion
Each slot in the `slots` section must include the `mappings` key. The same keys used for predefined mappings in 2.0 are available in 3.0. Additionally, you can define slots with
custom mappings implemented in a custom action, for example:
```

Author: Every slot that gets listed in the domain needs to have the `mappings` parameter defined, as well as everytime a `Slot` class gets instantiated (although I don't have real-life examples for the latter part). There is a little hack I used in the case of default slots internally, (e.g, `session_started_metadata`) meaning I assigned an empty list to mappings argument 😄 But that's not something I advise for domain slots. Thinking about this now, I should actually add a test case for this, when a domain slot has no mapping in `mappings`. I assume it would never get set when `action_extract_slots` runs.

Author: To follow-up on this, if a slot doesn't specify mappings (via the domain hack of `mappings: []`), that slot never gets set via a form or `action_extract_slots`.

Reviewer: No idea why I only saw this now, but cool! What is the difference between doing `mappings: []` and `mappings: - type: custom` without any custom action to extract the slot?

There should indeed be a test case for empty mappings.

Author: There's no difference really, other than that `type: custom` is an explicit way of indicating this slot can be set in other ways than the predefined ways (could be custom component or validation actions etc).

```yaml
slots:
  is_existing_customer:
    type: bool
    mappings:
    - type: custom
      action: action_verify_customer_status
```

> You can also create a custom validation action extending the `ValidationAction` in the Rasa SDK to either

Reviewer: Suggest adding a sub-heading under slot mappings for `slot validation methods` or something similar where this is covered.

> You can also use [slot validation actions](./slot-validation-actions.mdx) to define custom validation of slots with
> predefined mappings or custom extraction and validation of slots with custom mappings.

Reviewer: ```suggestion
You can also use [slot validation actions](./slot-validation-actions.mdx) to either validate slots with predefined mappings, or to both extract and validate slots with custom mappings.
```
If I'm reading it right, the point is that a slot validation action can be used for only validation, BUT only for slots with predefined mappings. To validate a slot with a custom mapping, both mapping and validation must sit in the same slot validation action. Suggestion is based on that assumption

:::note

> The `required_slots` of a form has been converted in a list of slot names because slot mappings are now relocated
> from `forms` to `slots` in the domain.

Reviewer: ```suggestion
The `required_slots` of a form used to be a list of slot mappings.
Since slot mappings are relocated to the `slots` section of the domain, `required_slots` has been converted to a list of slot names only.```

> need to list `action_validate_slot_mappings` in the domain `actions`.

Reviewer: This info should rather go in the doc on slot mappings, and the migration guide can refer to it

Because slot mappings are now relocated from `forms` to `slots`, the `required_slots` of a form has been converted in a

> list of slot names.

Reviewer: This point is very N.B. from an end user perspective. It should rather go first, then the explanation of where mappings have moved to.

Because slot mappings are now relocated from `forms` to `slots`, the `required_slots` of a form has been converted in a
list of slot names.

> 

Reviewer 2: I would add an extra header here to called something like #### Automatic slot migration from 2.0 to 3.0 to make it easier to skim

> To migrate automatically to the 3.0 domain format, you can run the following command:

Reviewer: Mentioned in the glossary too, but a blurb that only the domain file changes in format would be good

```bash
rasa data migrate -d DOMAIN --out OUT_PATH
```

In addition to creating a a valid 3.0 domain in the indicated out path, this command will automatically backup your
original domain file(s) in a file labeled `original_domain.yml` or `original_domain` directory if a directory was
provided instead.

> Please note that the migrated slot mappings from forms will include mapping conditions for each form.
> To learn more on mapping conditions, you can read more in this [section](./domain.mdx#mapping-conditions).

Reviewer: ```suggestion
To maintain the behavior of forms in the 2.0 format, all migrated slot mappings will include mapping conditions for each form. This can be changed manually according to your use case.
See the docs on [mapping conditions](./domain.mdx#mapping-conditions) for more information.
```

Since all migrated slot mappings will be coming from forms

#### Manually migrating from 2.0 domain format to the 3.0 format

> Let's consider the following 2.0 domain file:

Reviewer: First briefly list the steps for migration, then the worked example

This key is a list of mappings moved from forms, while the `required_slots` field collapses to a list of slot names.

Let's consider the following 2.0 domain file:

```yaml
entities:
 - cuisine
 - number
 slots:
   cuisine:
     type: text
   num_people:
     type: float
   outdoor_seating:
     type: bool
 forms:
   restaurant_form:
     required_slots:
         cuisine:
           - type: from_entity
             entity: cuisine
         num_people:
           - type: from_entity
             entity: number
         outdoor_seating:
           - type: from_intent
             intent: affirm
             value: true
           - type: from_intent
             intent: deny
             value: false
```

> The initial result of this migration could look like this:

Reviewer: ```suggestion
The initial result of migrating this domain to 3.0 format would look like this:
```

```yaml
entities:
 - cuisine
 - number
 slots:
   cuisine:
     type: text
     mappings:
     - type: from_entity
       entity: cuisine
   num_people:
     type: float
     mappings:
     - type: from_entity
       entity: number
   outdoor_seating:
     type: bool
     mappings:
     - type: from_intent
       intent: affirm
       value: true
     - type: from_intent
       intent: deny
       value: false
 forms:
   restaurant_form:

>      required_slots:
> 	 - cuisine
>      - num_people
>      - outdoor_seating

Reviewer: ```suggestion
     required_slots:
     - cuisine
     - num_people
     - outdoor_seating
```

```

> In case you'd prefer certain slots to be filled only within the context of a form, you can add mapping conditions like so:

Reviewer: ```suggestion
For slots that should be filled only in the context of a form, add mapping conditions to specify which form(s) should be active. Adding `conditions` is required preserve the behavior of slot mappings from 2.0, since without them the mappings will be applied on each user turn regardless of whether a form is active or not.
```

```yaml
 slots:
   outdoor_seating:
     type: bool
     mappings:
     - type: from_intent
       intent: affirm
       value: true
       conditions:
       - active_loop: restaurant_form
         requested_slot: outdoor_seating
     - type: from_intent
       intent: deny
       value: false
       conditions:
       - active_loop: restaurant_form
         requested_slot: outdoor_seating
```

Please note it is not required to use mapping conditions for custom slot mappings defined in a custom action inheriting
from `FormValidationAction`. However if you wish certain custom slot mappings to only be set within the context of a form,

> make sure your custom action inherits from `FormValidationAction` rather than from `ValidationAction`.

Reviewer: Keep explanation mostly in the main doc and link to it since the point here is domain format, not validation behaviour.

The `ConveRTTokenizer`, `LanguageModelTokenizer`, and `HFTransformersNLP` featurizer
components were deprecated in Rasa 2.x and have been removed in Rasa 3.0. See the
[migration guide for Rasa 2.x](#deprecations-2) for replacing these components in your pipeline.

### Slot Mappings

> As of version 3.0, there is a single explicit mechanism for slot filling enabled by defining slot mappings for each slot
> in the `slots` section of the domain file. This approach keeps slots up to date over the course of a conversation, and
> removes duplicated effort in mapping the same slots in multiple forms.
> 

Reviewer: Similar to the comment above about `custom` mappings being allowed to be empty: The phrasing here could make a user think they can't define "empty" slots that are filled only at specific points in the conversation anymore. But they can - they can just do more than that. 
```suggestion
As of version 3.0, there is a single explicit mechanism for slot filling enabled by defining slot mappings for each slot
in the `slots` section of the domain file. This approach keeps slots up to date over the course of a conversation, and
removes duplicated effort in mapping the same slots in multiple forms. It is still possible to fill slots from arbitrary custom actions and not update them on every turn of the conversation if that behavior is desired.

```

This new mechanism replaces the implicit slot setting via auto-fill of slots with entities of the same name.
The `auto_fill` key in the domain is no longer available, as well as the `auto_fill` parameter in the constructor of
the `Slot` class.

While forms continue to request the next slot, slot extraction is now delegated to the default
action [`action_extract_slots`](./default-actions.mdx#action_extract_slots). This action runs in the background
automatically after each user turn. Like `action_listen`, it should not be included in stories.

Each slot in the `slots` section must include the `mappings` key. The same keys used for predefined mappings in 2.0 are

> available in 3.0. Additionally, you can define slots with custom mappings implemented in a custom action, for example:
> 
> ```yaml
> slots:
>   is_existing_customer:
>     type: bool
>     mappings:
>     - type: custom
>       action: action_verify_customer_status
> ```
> 
> You can also use [slot validation actions](./slot-validation-actions.mdx) to either validate slots with predefined
> mappings, or to both extract and validate slots with custom mappings.

Reviewer: ```suggestion
available in 3.0. Additionally, you can define slots with custom mappings implemented in a custom action which will be run on every user turn, for example:

```yaml
slots:
  is_existing_customer:
    type: bool
    mappings:
    - type: custom
      action: action_verify_customer_status
\```

You can use [slot validation actions](./slot-validation-actions.mdx) to either validate slots with predefined
mappings, or to both extract and validate slots with custom mappings.

Slots which will be filled by arbitrary custom actions in the course of the conversation, and which should not be updated on every user turn, should be listed with mappings of type `custom` and no action. For example:

```yaml
slots:
  handoff_completed:
    type: bool
    initial_value: false
    mappings:
    - type: custom
\```

 This slots value will only change when a custom action is predicted that sets it. This mapping maintains the behavior from 2.x for a slot which was not filled by an entity or by slot mappings in a form.

```
